File: sdk/chocobase-scala/src/main/scala/io/chocobase/Postgrest.scala
package io.chocobase

class Postgrest(baseUrl: String, table: String, headers: Map[String, String], params: Map[String, String] = Map.empty) {
  def select(columns: String = "*"): Postgrest =
    new Postgrest(baseUrl, table, headers, params + ("select" -> columns))

  def eq(column: String, value: Any): Postgrest =
    new Postgrest(baseUrl, table, headers, params + (column -> s"eq.$value"))

  def limit(count: Int): Postgrest =
    new Postgrest(baseUrl, table, headers, params + ("limit" -> count.toString))

  def execute(): ujson.Value = {
    val queryStr = params.map { case (k, v) => s"$k=$v" }.mkString("&")
    val url = s"$baseUrl/rest/v1/$table" + (if (queryStr.nonEmpty) s"?$queryStr" else "")
    val resp = requests.get(url, headers = headers)
    ujson.read(resp.text())
  }
}

File: sdk/chocobase-kt/src/main/kotlin/io/chocobase/Postgrest.kt
package io.chocobase

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.OkHttpClient
import okhttp3.Request

class QueryBuilder(
    private val baseUrl: String,
    private val table: String,
    private val httpClient: OkHttpClient,
    private val headers: Map<String, String>
) {
    private val params = mutableMapOf<String, String>()
    private val gson = Gson()

    fun select(columns: String = "*"): QueryBuilder {
        params["select"] = columns
        return this
    }

    fun eq(column: String, value: Any): QueryBuilder {
        params[column] = "eq.$value"
        return this
    }

    fun limit(count: Int): QueryBuilder {
        params["limit"] = count.toString()
        return this
    }

    fun execute(): List<Map<String, Any>> {
        val urlBuilder = "$baseUrl/rest/v1/$table".toHttpUrlOrNull()?.newBuilder() ?: return emptyList()
        params.forEach { (k, v) -> urlBuilder.addQueryParameter(k, v) }

        val builder = Request.Builder().url(urlBuilder.build()).get()
        headers.forEach { (k, v) -> builder.addHeader(k, v) }

        httpClient.newCall(builder.build()).execute().use { response ->
            val body = response.body?.string() ?: "[]"
            val type = object : TypeToken<List<Map<String, Any>>>() {}.type
            return try {
                gson.fromJson(body, type)
            } catch (_: Exception) {
                emptyList()
            }
        }
    }
}
